internal/quickfort/dig.lua

```diff
@@ -1,26 +1,569 @@
+    return true
+end
+
+local function do_remove_designation(ctx)
+    if not has_designation(ctx.flags, ctx.occupancy) then return false end
+    clear_designation(ctx.flags, ctx.occupancy)
+    return true
+end
+
+local function is_valid_item(item)
+    return not item.flags.garbage_collect
+end
+
+local function get_items_at(pos, include_buildings)
+    local items = {}
+    if include_buildings then
+        local bld = dfhack.buildings.findAtTile(pos)
+        if bld and same_xyz(pos, xyz2pos(bld.centerx, bld.centery, bld.z)) then
+            for _, contained_item in ipairs(bld.contained_items) do
+                if is_valid_item(contained_item.item) then
+                    table.insert(items, contained_item.item)
+                end
+            end
+        end
+    end
+    for _, item_id in ipairs(dfhack.maps.getTileBlock(pos).items) do
+        local item = df.item.find(item_id)
+        if same_xyz(pos, item.pos) and
+                is_valid_item(item) and item.flags.on_ground then
+            table.insert(items, item)
+        end
+    end
+    return items
+end
+
+local function do_item_flag(pos, flag_name, flag_value, include_buildings)
+    local ret = false
+    for _, item in ipairs(get_items_at(pos, include_buildings)) do
+        item.flags[flag_name] = flag_value
+        ret = true
+    end
+    return ret
+end
+
+local function do_claim(ctx)
+    return do_item_flag(ctx.pos, "forbid", values.item_claimed, true)
+end
+
+local function do_forbid(ctx)
+    return do_item_flag(ctx.pos, "forbid", values.item_forbidden, true)
+end
+
+local function do_melt(ctx)
+    -- the game appears to autoremove the flag from unmeltable items, so we
+    -- don't actually need to do any filtering here
+    return do_item_flag(ctx.pos, "melt", values.item_melted, false)
+end
+
+local function do_remove_melt(ctx)
+    return do_item_flag(ctx.pos, "melt", values.item_unmelted, false)
+end
+
+local function do_dump(ctx)
+    return do_item_flag(ctx.pos, "dump", values.item_dumped, false)
+end
+
+local function do_remove_dump(ctx)
+    return do_item_flag(ctx.pos, "dump", values.item_undumped, false)
+end
+
+local function do_hide(ctx)
+    return do_item_flag(ctx.pos, "hidden", values.item_hidden, true)
+end
+
+local function do_unhide(ctx)
+    return do_item_flag(ctx.pos, "hidden", values.item_unhidden, true)
+end
+
+local function do_traffic_high(ctx)
+    if ctx.flags.hidden then return false end
+    ctx.flags.traffic = values.traffic_high
+end
+
+local function do_traffic_normal(ctx)
+    if ctx.flags.hidden then return false end
+    ctx.flags.traffic = values.traffic_normal
+end
+
+local function do_traffic_low(ctx)
+    if ctx.flags.hidden then return false end
+    ctx.flags.traffic = values.traffic_low
+end
+
+local function do_traffic_restricted(ctx)
+    if ctx.flags.hidden then return false end
+    ctx.flags.traffic = values.traffic_restricted
+end
+
+-- add on the 'd' prefix to avoid spelling out reserved words (like 'or')
```

Comment:
Lua has an alternative syntax for using arbitrary expressions as table keys: `{['do'] = 2}`
(You would also need to index with `table['do']` instead of `table.do`, but you're doing that already)

Reply:
Thanks! I should have remembered that from the Lua tutorial. done.